File: tutorial/2/lufact/F/lufact.f90
program lufact

  implicit none

  integer, parameter :: nmax = 100

  real, dimension(nmax,nmax) :: a, l, u, lu
  real, dimension(nmax)   :: x, b, ax

  integer :: i, j

  integer :: n = 4

  call matgen(a,nmax,n,b)

  write(*,*)
  write(*,*) 'a'
  write(*,*)
  call matprint(a, nmax, n)

  write(*,*)
  write(*,*) 'b'
  write(*,*)
  call vecprint(b,n)

  write(*,*)
  write(*,*) 'Need to perform LU factorisation of A'
  write(*,*) 'Currently setting L=U=0'

!
! Set l and u to zero
!

  do i = 1, n
    do j = 1, n
      l(i,j) = 0.0
      u(i,j) = 0.0
    end do
  end do

  write(*,*)
  write(*,*) 'l'
  write(*,*)
  call matprint(l, nmax, n)


  write(*,*)
  write(*,*) 'u'
  write(*,*)
  call matprint(u, nmax, n)

!
!  Multiply the L and U factors together
!

  call matmat(lu, l, u, nmax, n)

  write(*,*)
  write(*,*) 'lu'
  write(*,*)
  call matprint(lu, nmax, n)


  write(*,*)
  write(*,*) 'Need to forward and back substitution to solve Ax=b'
  write(*,*) 'Currently setting x=b'

  do i = 1, n
    x(i) = b(i)
  end do


  write(*,*)
  write(*,*) 'x'
  write(*,*)
  call vecprint(x,n)

!
! Set y = A x and print out
!

  call matvec(ax, a, nmax, x, n)

  write(*,*)
  write(*,*) 'Ax'
  write(*,*)
  call vecprint(ax,n)

end program lufact


!  matprint
!
!  Prints a matrix of reals to the screen. Matrix is taken to be
!  of size n x n although it must have been defined with a leading
!  dimension of lda.


subroutine matprint(a, lda, n)

  implicit none

  integer :: lda, n, i, j

  real, dimension(lda,n) :: a

  if (n .gt. 6) then
    write(*,*) 'Matrix too large to print'
  else

    do i = 1, n
      write(*,10) (a(i,j), j=1,n)
10    format(10(1x, f12.8))
    end do

  end if

end subroutine matprint


!  vecprint
!
!  Prints a vector containing n reals to the screen.

subroutine vecprint(x, n)

  implicit none

  integer :: n, i

  real, dimension(n) :: x

  if (n .gt. 6) then
    write(*,*) 'Vector too large to print'
  else

    do i = 1, n
      write(*,*) x(i)
10    format(1x, f12.8)
    end do

  end if

end subroutine vecprint

!  matvec
!
!  Sets y = a.x, where y and x are vectors of reals of length n, and a
!  is an n x n matrix. Matrix must have been defined with a leading
!  dimension of lda.


subroutine matvec(y, a, lda, x, n)

  implicit none

  integer :: lda, n, i, j

  real, dimension(n)     :: x, y
  real, dimension(lda,n) :: a

  do i = 1, n
    y(i) = 0.0

    do j = 1, n
      y(i) = y(i) + a(i,j)*x(j)
    end do

  end do

end subroutine matvec


!  matmat
!
!  Sets a = b.c, where a, b and c are all n x n matrices of reals.
!  All matrices must have been defined with a leading dimension lda.

subroutine matmat(a, b, c, lda, n)

  implicit none

  integer :: lda, n, j

  real, dimension(lda,n) :: a, b, c

  do j = 1, n

    call matvec(a(1,j), b, lda, c(1,j), n)

  end do

end subroutine matmat


!  matgen
!
!  Constructs a random matrix of size n x n, and a solution vector b
!  Matrix must have been defined with a leading dimension lda.


subroutine matgen(a,lda,n,b)

  implicit none

  integer                :: lda, n
  real, dimension(n)     :: b
  real, dimension(lda,n) :: a

  integer :: i, j, init

  init = 1325;

  do j = 1, n
    do i = 1, n

       init = mod(3125*init, 65536)
       a(i,j) = (init - 32768.0)/16384.0

    end do
  end do

  b(:) = sum(a, dim=2)

end subroutine matgen

File: tutorial/2/lufact/C/lufact.c
#include <stdio.h>
#include <math.h>

#include <LU.h>

#define NMAX 100

void matprint(int n, float a[NMAX][NMAX]);
void vecprint(int n, float x[NMAX]);
void matvec(int n, float y[NMAX], float a[NMAX][NMAX], float x[NMAX]);
void matmat(int n, float a[NMAX][NMAX], float b[NMAX][NMAX], float c[NMAX][NMAX]);
void matgen(int n, float a[NMAX][NMAX], float b[NMAX]);

// My implementation:
void identificate(int n, float M[NMAX][NMAX]);
void setZero(int n, float M[NMAX][NMAX]);
void LUfact(int n, float A[NMAX][NMAX], float L[NMAX][NMAX], float U[NMAX][NMAX]);


void main(void)
{
  float a[NMAX][NMAX], l[NMAX][NMAX], u[NMAX][NMAX], lu[NMAX][NMAX];
  float x[NMAX], b[NMAX], ax[NMAX];
  float y[NMAX];

  int i, j;

  int n = 4;  

  matgen(n, a, b);

  printf("\na\n\n");
  matprint(n, a);

  printf("\nb\n\n");
  vecprint(n, b);

  printf("\nNeed to perform LU factorisation of A\n");
  printf("Currently setting L=U=0\n");

  /*
   * Set l and u to zero
   */

  for (i=0; i < n; i++)
  {
    for (j=0; j < n; j++)
    {
      l[i][j] = 0.0;
      u[i][j] = 0.0;
    }
  }

  // ---------------------------------------------------my implementation-----------------------------
  identificate(n, l);
  LUfact(n, a, l, u);

  printf("\nl\n\n");
  matprint(n, l);

  printf("\nu\n\n");
  matprint(n, u);

  /*
   * Multiply the L and U factors together
   */

  matmat(n, lu, l, u);

  printf("\nlu\n\n");
  matprint(n, lu);
  
  printf("\nNeed to forward and back substitution to solve Ax=b\n");
  printf("Currently setting x=b\n");

  for (i=0; i < n; i++)
  {
    x[i] = b[i];
  }

  printf("\nx\n\n");
  vecprint(n, x);
  
  /*
   * Set y = A x and print out
   */

  matvec(n, ax, a, x);

  printf("\nAx\n\n");
  vecprint(n, ax);
}


/*  matprint
 *
 *  Prints a matrix of floats to the screen. Matrix is taken to be
 *  of size n x n although it must have been defined as size NMAX x NMAX
 *  where NMAX is a compile-time constant.
 */


void matprint(int n, float a[NMAX][NMAX])
{
  int i, j;

  if (n > 6)
  {
    printf("Matrix too large to print\n");
  }
  else
  {
    for (i=0; i < n; i++)
    {
      for (j=0; j < n; j++)
      {
        printf(" %12.8f", a[i][j]);
      }
      printf("\n");
    }
  }
}


/*  vecprint
 *
 *  Prints a vector of floats to the screen. Vector is taken to be
 *  of size n although it must have been defined as size NMAX
 *  where NMAX is a compile-time constant.
 */


void vecprint(int n, float x[NMAX])
{
  int i;

  if (n > 6)
  {
    printf("Vector too large to print\n");
  }
  else
  {
    for (i=0; i < n; i++)
    {
      printf(" %12.8f\n", x[i]);
    }
  }
}


/*  matvec
 *
 *  Sets y = a.x, where y and x are vectors of floats of length n, and a
 *  is an n x n matrix. All arrays must have been defined as size NMAX
 *  where NMAX is a compile-time constant.
 */

void matvec(int n, float y[NMAX], float a[NMAX][NMAX], float x[NMAX])
{
  int i, j;

  for (i=0; i < n; i++)
  {
    y[i] = 0.0;

    for (j=0; j < n; j++)
    {
      y[i] += a[i][j]*x[j];
    }
  }
}


/*  matmat
 *
 *  Sets a = b.c, where a, b and c are all n x n matrices of floats.
 *  All matrices must have been defined as size NMAX x NMAX where NMAX
 *  is a compile-time constant.
 */


void matmat(int n, float a[NMAX][NMAX], float b[NMAX][NMAX], float c[NMAX][NMAX])
{
  int i, j, k;

  for (i=0; i < n; i++)
  {
    for (j=0; j < n; j++)
    {
      a[i][j] = 0.0;

      for (k=0; k < n; k++)
      {
        a[i][j] += b[i][k]*c[k][j];
      }
    }
  }
}


/*  matgen
 *
 *  Constructs a random matrix of size n x n, and a solution vector b
 *  All arrays must have been defined as size NMAX where NMAX is a
 *  compile-time constant.
 */

void matgen(int n, float a[NMAX][NMAX], float b[NMAX])
{
	int init, i, j;

	init = 1325;
	for (j = 0; j < n; j++) {
		for (i = 0; i < n; i++) {
			init = 3125*init % 65536;
			a[i][j] = (init - 32768.0)/16384.0;
		}
	}
	for (i = 0; i < n; i++) {
          b[i] = 0.0;
	}
	for (j = 0; j < n; j++) {
		for (i = 0; i < n; i++) {
			b[i] = b[i] + a[i][j];
		}
	}
}


void identificate(int n, float M[NMAX][NMAX]) {
    for(int i=0;i<n;i++){
        for(int j=0;j<n;j++){
            if(i==j)
                M[i][j]=1.0;
        }
    }
}

void setZero(int n, float M[NMAX][NMAX]) {
    for(int i=0;i<n;i++){
        for(int j=0;j<n;j++){
                M[i][j]=0.0;
        }
    }
}

void LUfact(int n, A[NMAX][NMAX], L[NMAX][NMAX], U[NMAX][NMAX]) {
    for(int j=0;j<n;j++){
        for(int i=0;i<j;i++){
            float sum=0.0;
            for(int k=0;k<i-1;k++){
                sum+=(L[i][k]*U[k][j])
            }
            U[i][j]=A[i][j]-sum;
        }
        for(int i=j;i<n;i++){
            float sum=0.0;
            for(int k=0;k<j-1;k++){
                sum+=(L[i][k]*U[k][j])
            }
            L[i][j]=(a[i][j]-sum)/U[i][j];
        }
    }
}